src/app/rpmostree-composeutil.c

```diff
@@ -468,3 +468,169 @@ rpmostree_composeutil_write_composejson (OstreeRepo  *repo,
+
+      g_variant_builder_close (&builder); /* $repoid */
+    }
+  g_variant_builder_close (&builder); /* "packages" */
+
+  if (path)
+    {
+      g_autoptr(GVariant) lock_v = g_variant_builder_end (&builder);
+      g_assert (lock_v != NULL);
+      g_autoptr(JsonNode) lock_node = json_gvariant_serialize (lock_v);
+      g_assert (lock_node != NULL);
+      glnx_unref_object JsonGenerator *generator = json_generator_new ();
+      json_generator_set_root (generator, lock_node);
+
+      char *dnbuf = strdupa (path);
+      const char *dn = dirname (dnbuf);
+      g_auto(GLnxTmpfile) tmpf = { 0, };
+      if (!glnx_open_tmpfile_linkable_at (AT_FDCWD, dn, O_WRONLY | O_CLOEXEC, &tmpf, error))
+        return FALSE;
+      g_autoptr(GOutputStream) out = g_unix_output_stream_new (tmpf.fd, FALSE);
+      /* See also similar code in status.c */
+      if (json_generator_to_stream (generator, out, NULL, error) <= 0 ||
+          (error != NULL && *error != NULL))
+        return FALSE;
+
+      /* World readable to match --write-commitid-to which uses mask */
+      if (!glnx_fchmod (tmpf.fd, 0644, error))
+        return FALSE;
+
+      if (!glnx_link_tmpfile_at (&tmpf, GLNX_LINK_TMPFILE_REPLACE, AT_FDCWD, path, error))
+        return FALSE;
+
+      g_print("Saved lockfile to %s\n", path);
+    }
+
+  return TRUE;
+}
+
+/** FIXME **/
+/* compose tree accepts JSON package version lock via file;
+ * convert it to a hash table of a{sv}; suitable for further extension.
+ */
+GHashTable *
+rpmostree_composeutil_get_vlockmap (JsonParser *parser,
+                                    GError    **error)
+{
+  g_autoptr(GHashTable) name_to_nevra =
+    g_hash_table_new_full (g_str_hash, g_str_equal, g_free, g_free);
+
+  /* lockfile = {sa{sa(ss)}} */
+  JsonNode *metarootval = json_parser_get_root (parser);
+  g_autoptr(GVariant) jsonmetav = json_gvariant_deserialize (metarootval, "{sa{sa(ss)}}", error);
+  if (!jsonmetav)
+    return NULL;
+
+  const char *pkgkey;
+  g_autoptr(GVariant) repometav = NULL;
+  g_variant_get (jsonmetav, "{&s@a{sa(ss)}}", &pkgkey, &repometav);
+  if (strcmp (pkgkey, "packages") != 0)
+    {
+      g_set_error (error, G_IO_ERROR, G_IO_ERROR_NOT_FOUND,
+                   "Failed to find \"packages\" section in lock file");
+      return NULL;
+    }
+
+  if (!repometav)
+    return NULL;
+
+  char *repo;
+  GVariant *value;
+  GVariantIter viter;
+  g_variant_iter_init (&viter, repometav);
+  while (g_variant_iter_loop (&viter, "{&s@a(ss)}", &repo, &value))
+    {
+      GVariantIter piter;
+      g_variant_iter_init (&piter, value);
+      char *nevra, *repochksum;
+      while (g_variant_iter_loop (&piter, "(s&s)", &nevra, &repochksum))
+        {
+          char *name = NULL;
+          if (!rpmostree_decompose_nevra (nevra, &name, NULL, NULL, NULL, NULL, error))
+            return NULL;
+          g_hash_table_insert (name_to_nevra, name, g_strdup (nevra));
+          /* FIXME: how to return repochksum info? */
```

Comment:
Hmm, let's just return a hash table of `(nevra -> checksum)` here? (See related comment below.)

Reply:
Resolved